cmd/erasure-healfile.go

```diff
@@ -36,48 +50,72 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for blockOffset = 0; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			blocks[i] = make([]byte, chunksize)
```

Comment:
```
if blocks[i] == nil || len(blocks[i]) < len(chunksize) {
     blocks[i] = make([]byte, chunksize)
}
```
To avoid unnecessary memory alloc.

Reply:
`blocks[i] == nil` is guaranteed just before this line, so I don't see why your suggestion reduces allocation. However, I am updating the PR with a better way to reduce allocations.

Reply:
Yes, but if one read fails (let's say one disk goes down while healing) the buffer would be set to nil again. But your updated solution is also fine. Both are equivalent.

Reply:
Not sure if your last comment means you are happy with the allocations behaviour - I think it is ok. My thoughts about further optimization for this are here - https://github.com/minio/minio/commit/f28e723ea82c222bda62549857746a97c277041c#commitcomment-24306122.